Messaging.md

```diff
@@ -44,20 +55,114 @@ The following convenience types are also defined:
+* `x_point`: a 32-byte x-only public key with implicit y-coordinate being even as in [BIP 340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#design)
 * `point`: a 33-byte Elliptic Curve point (compressed encoding as per [SEC 1 standard](http://www.secg.org/sec1-v2.pdf#subsubsection.2.3.3))
 * `spk`: A bitcoin script public key encoded as ASM prefixed with a Bitcoin CompactSize unsigned integer
 * `short_contract_id`: an 8 byte value identifying a contract funding transaction on-chain (see [BOLT #7](https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#definition-of-short-channel-id))
 * `bigsize`: a variable-length, unsigned integer similar to Bitcoin's CompactSize encoding, but big-endian.  Described in [BigSize](https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md#appendix-a-bigsize-test-vectors).
-* `contract_info`: ???
-* `oracle_info`: ???
-* `funding_input`: ???
-* `cet_signatures`: ???
-* `funding_signatures`: ???
+
+## DLC Specific Types
+
+The following DLC-specific types are used throughout the specification. All type numbers are placeholders subject to change both here and in [Protocol.md](Protocol.md).
+
+### The `contract_info` Type
+
+This type contains information about a contracts outcomes and their corresponding payouts. To save space, only one side's POV is included in this message as the other can be derived using `remote_payout = total_collateral - local_payout`.
+
+#### Version 0 `contract_info`
+
+1. type: 42768 (`contract_info_v0`)
+2. data:
+   * [`sha256`:`outcome_1`]
+   * [`u64`:`outcome_1_local_payout`]
+   * ...
+   * [`sha256`:`outcome_n`]
+   * [`u64`:`outcome_n_local_payout`]
+
+This type of contract info is a simple enumeration where the value `n` is omitted from being explicitly included as it can be derived from the length field of the TLV.
+
+### The `oracle_info` Type
+
+This type contains information about the oracle(s) to be used in executing a DLC, and possibly the outcomes possible if these are not specified in the corresponding `contract_info`.
+
+#### Version 0 `oracle_info`
+
+1. type: 42770 (`oracle_info_v0`)
+2. data:
+   * [`x_point`:`oracle_public_key`]
+   * [`x_point`:`oracle_nonce`]
+
+This type of oracle info is for single-oracle, single signature (and hence single nonce) events.
+
+### The `funding_input` Type
+
+This type contains information about a specific input to be used in a funding transaction, as well as its corresponding on-chain UTXO.
+
+#### Version 0  `funding_input`
+
+1. type: 42772 (`funding_input_v0`)
+2. data:
+   * [`u16`:`prevtx_len`]
+   * [`prevtx_len*byte`:`prevtx`]
+   * [`u32`:`prevtx_vout`]
+   * [`u32`:`sequence`]
+   * [`u16`:`max_witness_len`]
+   * [`u16`:`redeemscript_len`]
+   * [`redeemscript_len*byte`:`script`]
+
+`prevtx_tx` is the serialized transaction whose `prevtx_vout` output is being spent.
```

Comment:
This is a deal breaker for me. 

In BTCPay Server, we don't have the prevtx, because if the user restored his wallet, it is through scantxoutsetinfo and we work on pruned node.

Reply:
I think this requirement can be dropped, because the signatures already cover the input value (it's a segwit feature).

Reply:
Well it was not needed in the non-TLV version of the spec, why has it been added?

Reply:
@devrandom it covers your own input value but not the others https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification

If you don't include the full prev tx they can give you a valid txid pointing to a non-segwit address and convince you it is segwit (because you're trusting them by not requiring the full prevtx) once you sign the funding tx, they can include a script signature you weren't expecting, your signatures are still valid but the txid changes and all CETs and refund tx become invalid.

Reply:
It wasn't there before because I was lazy, they should be there, they also should be present in PSBTs as well as that they are in the LN dual funding proposal

Reply:
#TaprootFixesThis because then the signatures will commit to the other inputs' spk

Reply:
Re ariard I agree, in bitcoin-s we had to add a transaction table to our db where we keep track of the prevtxs of our current UTXO set

Reply:
@nkohen even with this, the previous transactions should not be required. Let me explain why it is problematic, and why it is not needed.

Users restoring wallet on BTCPay Server, are using `scantxoutset` RPC command. This allow them to restore the balance of a wallet, without having to redownload blocks.

The downside of it, is that our users don't have access to their previous transaction. And even if they do have access, this is really not user friendly as, when they create the PSBT, they need to specify that they need to include the previous TX in the PSBT. (A normal user have no idea what it means, and would be confused if it does not work when he needs it)

So basically, as far as I am concerned, requiring the previous tx as part of the protocol either means:
1. I stop working on DLC until taproot fixes this
2. I use a variant of this protocol for NDLC
3. This protocol revert this change and does not requires the previous transaction.

Now, let me explain why `3.` does not represent a security risk.

In NDLC, the signature of the funding transaction is done at [this step](https://github.com/dgarage/NDLC/blob/master/docs/Concepts.md#alice-starts-the-contract), and it can be done with any wallet supporting PSBT.

Given the attack you explained, NDLC will not include the witness_utxo or non_witness_utxo in the funding PSBT that it asks the user to sign.

All wallet signing PSBTs, in such condition, can't sign without updating the PSBT with data from their own UTXO Set, meaning that the witness_utxo is not controlled by the attacker.
On top of it, I will require, at NDLC level, that when the user submit the partially signed PSBT of the funding transaction, the witness_utxo is properly filled. This would detect any wallet blindly signing and thus being a danger for the user.

Such PSBT update can be trivially done by bitcoin core using only information from the UTXO Set (and thus, by BTCPay, which has access to it).
I believe it can also be done by Electrum. (Yes, trusting the electrum server)

Reply:
my attack has `witness_utxo` set but that field only has an output and that output isn't committed to by your signature (so the coutner-party can use something else)

Reply:
> my attack has witness_utxo set but that field only has an output and that output isn't committed to by your signature (so the counter-party can use something else)

The counter-party can't use another output, as the outpoint is committed to. (by `hashPrevouts`)

However, he can trick me by saying his input is segwit.
But my point is that somebody with the UTXO Set can protect against such attack without the need of the previous transaction.

However, for wallets without UTXO set, they can't verify it (Bitcoin-S and Wasabi wallet for example)

AFAIK, there is no UX available in bitcoin core to force the inclusion of previous transaction in PSBTs either.
So it means, you can't make DLC without having to develop a wallet from scratch for it. (and requiring full block scan rather than UTXO scan for restore)

I think this is a major hurdle, so on my side I will probably wait until taproot fix this before continuing my work on DLC, or use another protocol until then. I don't think there is a solution to solve this problem.

Reply:
So I agree with most of this but it isn't actually a problem if you don't have the UTXO set if they share the prevtx because worst case the prevtx is not actually real or the txo is already spent, in either case the funding tx will be invalid. This is ok. What is not okay is if you don't use prevtx, don't have the utxo set and they lie about there being a real txid, vout corresponding to an output (and they fake the output telling you it is segwit when it isn't) in this case they can lock up your funds.

Here is my proposal: The only safe way for light clients to do this is to use prevtx, otherwise they are trusting their coutner-party not to cheat them in this specific way which can lock up their funds. If a party has access to the UTXO set however, they really don't need or want prev-tx, so in the future I want there to be TLV versions of offer and accept that both have and don't have prev-tx and full nodes will receive non-prevtx messages while light clients will need prevtx. Because this is my plan, you will still be able to use the non-prevtx version (but it will be trusting) until you either have the infrastructure to support prevtx or taproot comes around after which this won't be necessary anymore. Thoughts?

Reply:
Also, the kind of good news is if your tool is somewhat manual, users can use external sources to validate prev_txid is in the utxo set so that it isn't counter-party trust

Reply:
oh and I think I misunderstood, if I now understand correctly, with this new versioned approach you should be able to be trustless if your counter-party accepts txid+vout instead of prevtx (if they are a full node for example) as the issue for you is getting your own prevtx, not validating the counter party's if you are given a prevtx right?

Reply:
@nkohen so I searched a bit:

* Bitcoin core is including previous transaction when creating a new PSBT for the wallet since 0.20.1. (latest version)
* So I swtiched BTCPay Server to mirror the same behavior by default. Sadly, this won't work for restored wallet.

> as the issue for you is getting your own prevtx, not validating the counter party's if you are given a prevtx right?

Correct.

Stepping back a bit, I think the best reason for allowing witness_utxo in the protocol, is not about making it easier for wallets with the full utxo set. But to prepare taproot. Once we have taproot, the non_witness_utxo won't be needed in any case. So I think it makes sense to allow it in the protocol for this reason.

Reply:
Great news :) Lmk if there's anything I can do to help. And also to clarify, you're cool with moving forward with prevtx and adding a non-prevtx version in the near future (potentially the subsequent PR to this one)?

Reply:
yes, seems good.

Reply:
https://github.com/discreetlogcontracts/dlcspecs/issues/98